File: src/ml-dotnet-book/src/digits-recognizer/DigitsRecognizer/EntryPoint.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using Humanizer;

namespace MachineLearningBook.DigitsRecognizer
{
    public static class EntryPoint
    {        
        public static void Main(string[] args)
        {
            // Path selection is very brittle due to the assumed structure.  If not just running for
            // personal playtime purposes, these should be passed in or otherwise set in configuration.

            var dataSetFile        = "digits.csv";
            var rootPath           = $"{ Assembly.GetEntryAssembly().Location }/../../.././../../../../data";
            var trainingPath       = Path.GetFullPath(Path.Combine(rootPath, $"training/{ dataSetFile }"));
            var validationPath     = Path.GetFullPath(Path.Combine(rootPath, $"validation/{ dataSetFile }"));
            var watch              = new Stopwatch();
            var distanceCalculator = (ImageDistanceCalculator<int>)ImageDistanceCalculators.Manhattan;
            var classifier         = new BasicImageClassifier(distanceCalculator);  

            Console.WriteLine();
            Console.WriteLine("Run started using:");            
            Console.WriteLine($"\tTraining Data Path: { trainingPath }");
            Console.WriteLine($"\tValidation Data Path: { validationPath }");
            Console.WriteLine($"\tClassifier: { classifier.GetType().Name.Humanize(LetterCasing.Title) } with { distanceCalculator.Method.Name } distance algorithm");
            Console.WriteLine();

            IReadOnlyList<ImageObservation<int>> ReadFileObservations(string path) =>
                File.ReadAllLines(trainingPath)
                    .Skip(1)
                    .Select(line => line.Split(',', StringSplitOptions.RemoveEmptyEntries))                    
                    .Select(tokens => new ImageObservation<int>(tokens[0], tokens.Skip(1).Select(token => Int32.Parse(token))))
                    .ToList();                              

            watch.Start();
            Console.Write("Preparing data...");
                    
            var trainingData   = ReadFileObservations(trainingPath);            
            var validationData = ReadFileObservations(validationPath);   

            watch.Stop();
            Console.WriteLine($" complete.  Elapsed: { watch.Elapsed.Humanize() }");

            watch.Reset();
            watch.Start();
            Console.Write("Training...");            
            
            classifier.Train(trainingData);

            watch.Stop();
            Console.WriteLine($" complete.  Elapsed: { watch.Elapsed.Humanize() }");

            watch.Reset();
            watch.Start();
            Console.Write($"Classifying...");
            
            var correctPercent = ImageClassifierEvaluator.CalculateCorrectPercentage<int>(validationData, classifier);

            watch.Stop();
            Console.WriteLine($" complete.  Elapsed: { watch.Elapsed.Humanize() }");

            Console.WriteLine();
            Console.WriteLine("Run complete.  {0:P2} classifications were correct.", correctPercent);
            Console.WriteLine();

            Console.Write("Press any key");
            Console.ReadKey();
        }
    }
}

File: src/ml-dotnet-book/src/spam-detector/SpamDetector/EntryPoint.fs
namespace MachineLearningBook.SpamDetector
    module EntryPoint =

        open System
        open System.Diagnostics
        open System.Reflection
        open Humanizer
        open System.IO
        open Classifier


        [<EntryPoint>]
        let Main argv =
            let dataSetFile    = "sms-spam.txt"
            let rootPath       = sprintf "%s/../../.././../../../../data" (Assembly.GetEntryAssembly().Location)
            let trainingPath   = Path.GetFullPath(Path.Combine(rootPath, "training/" + dataSetFile))
            let validationPath = Path.GetFullPath(Path.Combine(rootPath, "validation/" + dataSetFile))
            let runWatch       = new Stopwatch()
            let watch          = new Stopwatch()
        
            printfn ""
            printfn "Run started using:"
            printfn "\tTraining Data Path: %s" trainingPath
            printfn "\tValidaiton Data Path: %s" validationPath 
            printfn ""
        
            runWatch.Start()
            watch.Start()
            printf "Preparing data..."

            let trainingSet = 
                trainingPath 
                |> File.ReadAllLines 
                |> DataSet.parseData

            let validationSet = 
                validationPath 
                |> File.ReadAllLines 
                |> DataSet.parseData
        
            watch.Stop();
            printfn " complete.  Elapsed: %s" (watch.Elapsed.Humanize())
        
            watch.Reset()
            watch.Start()
            printf "Training..."

            let classifier = NaiveBayes.train trainingSet Tokenizer.wordBreakTokenizer (Set.empty.Add "txt")
        
            watch.Stop();
            printfn " complete.  Elapsed: %s" (watch.Elapsed.Humanize())

            watch.Reset()
            watch.Start()
            printf "Classifying..."

            let avgBy (docType, text) =
                match (classifier text) with
                | Some result when docType = result -> 1.0
                | _                                 -> 0.0

            let correct =
                validationSet
                |> Seq.averageBy avgBy
        
            watch.Stop();
            printfn " complete.  Elapsed: %s" (watch.Elapsed.Humanize())

            printfn ""
            printfn "%.3f%% of classifications were correct." correct

            runWatch.Stop()
            printfn "Run complete.  Total time: %s" (runWatch.Elapsed.Humanize())            
            
            printfn ""
            printfn ""
            printf "Press any key"
            Console.ReadKey true |> ignore
            0 
